src/Http/Http/src/Features/RequestBodyPipeFeature.cs

```diff
@@ -31,10 +33,13 @@ public PipeReader Reader
                 }
 
                 if (_internalPipeReader == null ||
-                    !object.ReferenceEquals(_internalPipeReader.InnerStream, _context.Request.Body))
+                    !ReferenceEquals(_streamInstanceWhenWrapped, _context.Request.Body))
                 {
-                    _internalPipeReader = new StreamPipeReader(_context.Request.Body);
-                    _context.Response.RegisterForDispose(_internalPipeReader);
+                    _streamInstanceWhenWrapped = _context.Request.Body;
+                    _internalPipeReader = PipeReader.Create(_context.Request.Body);
```

Comment:
@davidfowl the new wrappers aren't disposable, correct?

Reply:
Yeah, that's why the question. They aren't disposable but they seem to have resources that are cleaned up in `Complete`: https://github.com/dotnet/corefx/blob/master/src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeWriter.cs#L206

Do we need to make sure `Complete` is called?

Reply:
I think we really should make sure Complete is called... maybe adding it to OnCompleted as a callback? Either way I hate it.

@davidfowl what should we do here?

Reply:
OnCompleted

Reply:
Which `OnCompleted`? `HttpResponse.OnCompleted`?

Reply:
Is `Complete` idempotent?

Reply:
Internally, `Complete` is called only once (bool to check if it has been called already).